### PR TITLE
Generate rpm dependencies

### LINKED ISSUE #17142: Missing dependencies in rpm package
https://github.com/Microsoft/vscode/issues/13089 but for .rpm

Notes:

- Chromium deps: https://cs.chromium.org/chromium/src/chrome/installer/linux/rpm/expected_deps_i386?dr=C&q=libxss.so.1&sq=package:chromium
- `ldd /usr/share/code-insiders/code-insiders` can fetch a list of deps

### PR DESCRIPTION
**Pull request by rzhao271**
This PR fixes https://github.com/microsoft/vscode/issues/17142.

It's still in a real WIP state, though.
Based off of https://source.chromium.org/chromium/chromium/src/+/main:chrome/installer/linux/rpm/.

Questions:
1. Do we need to port the update script?
2. Is the requires field really supposed to have such a long list?

A screenshot showing the requires field of the rpm package spec with an auto-generated long list of dependencies

### CODE REVIEW

File: build/linux/rpmDependenciesGenerator.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+    // Get the files for which we want to find dependencies.
+    const findResult = (0, child_process_1.spawnSync)('find', ['.', '-name', '*.node']);
+    if (findResult.status) {
+        console.error('Error finding files:');
+        console.error(findResult.stderr.toString('utf-8'));
+        return [];
+    }
+    // Filter the files and add on the Code binary.
+    const files = findResult.stdout.toString('utf-8').split('\n').filter((file) => {
+        return !file.includes('os.target') && file.includes('build/Release');
```
[the remark this answers is not in the document]
Comment (rzhao271) [reply build/linux/rpmDependenciesGenerator.js:20]: Actually, I just realized this is the js file. The JS files shouldn't be in the PR, should they?

**Comment by ferdnyc** [comment]
To answer the second question, it is not outside the realm of possibility. Here's the requires list for `gnome-shell-41.4-1.fc35.x86_64`:

```sh
$ rpm -q --requires gnome-shell
/usr/bin/python3
/usr/bin/sh
accountsservice-libs(x86-64)
at-spi2-atk(x86-64)
bolt(x86-64)
control-center
gdm-libs(x86-64)
geoclue2-libs(x86-64)
gjs(x86-64) >= 1.69.2
glib2(x86-64) >= 2.56.0
gnome-bluetooth(x86-64) >= 1:3.9.0
gnome-desktop3(x86-64) >= 3.35.91
gnome-settings-daemon(x86-64) >= 3.37.1
gobject-introspection(x86-64) >= 1.49.1
gsettings-desktop-schemas(x86-64) >= 41~alpha
gstreamer1(x86-64) >= 1.4.5
gstreamer1-plugins-good(x86-64)
gtk3(x86-64) >= 3.15.0
gtk4(x86-64) >= 4.0.0
highcontrast-icon-theme
ibus(x86-64) >= 1.5.2
libGLESv2.so.2()(64bit)
libX11.so.6()(64bit)
libXfixes.so.3()(64bit)
libatk-1.0.so.0()(64bit)
libatk-bridge-2.0.so.0()(64bit)
libc.so.6()(64bit)
libc.so.6(GLIBC_2.14)(64bit)
libc.so.6(GLIBC_2.2.5)(64bit)
libc.so.6(GLIBC_2.28)(64bit)
libc.so.6(GLIBC_2.3)(64bit)
libc.so.6(GLIBC_2.33)(64bit)
libc.so.6(GLIBC_2.34)(64bit)
libc.so.6(GLIBC_2.4)(64bit)
libcairo.so.2()(64bit)
libecal-2.0.so.1()(64bit)
libedataserver-1.2.so.26()(64bit)
libgcc_s.so.1()(64bit)
libgcc_s.so.1(GCC_3.0)(64bit)
libgcc_s.so.1(GCC_3.3.1)(64bit)
libgcr-base-3.so.1()(64bit)
libgdk-3.so.0()(64bit)
libgdk_pixbuf-2.0.so.0()(64bit)
libgio-2.0.so.0()(64bit)
libgirepository-1.0.so.1()(64bit)
libgjs.so.0()(64bit)
libglib-2.0.so.0()(64bit)
libgnome-autoar-0.so.0()(64bit)
libgnome-desktop-3.so.19()(64bit)
libgnome-shell-menu.so()(64bit)
libgnome-shell.so()(64bit)
libgnomekbd
libgobject-2.0.so.0()(64bit)
libgraphene-1.0.so.0()(64bit)
libgtk-3.so.0()(64bit)
libgtk-4.so.1()(64bit)
libgweather(x86-64) >= 40~alpha
libical-glib.so.3()(64bit)
libjson-glib-1.0.so.0()(64bit)
libjson-glib-1.0.so.0(libjson-glib-1.0.so.0)(64bit)
libm.so.6()(64bit)
libm.so.6(GLIBC_2.2.5)(64bit)
libm.so.6(GLIBC_2.29)(64bit)
libmutter-9.so.0()(64bit)
libmutter-clutter-9.so.0()(64bit)
libmutter-cogl-9.so.0()(64bit)
libmutter-cogl-pango-9.so.0()(64bit)
libnm.so.0()(64bit)
libnm.so.0(libnm_1_0_0)(64bit)
libnm.so.0(libnm_1_4_0)(64bit)
libnma(x86-64)
libpango-1.0.so.0()(64bit)
libpangocairo-1.0.so.0()(64bit)
libpolkit-agent-1.so.0()(64bit)
libpolkit-gobject-1.so.0()(64bit)
libpulse-mainloop-glib.so.0()(64bit)
libpulse-mainloop-glib.so.0(PULSE_0)(64bit)
libpulse.so.0()(64bit)
libpulse.so.0(PULSE_0)(64bit)
librsvg2(x86-64)
libsecret-1.so.0()(64bit)
libst-1.0.so()(64bit)
libsystemd.so.0()(64bit)
libsystemd.so.0(LIBSYSTEMD_209)(64bit)
libwayland-server.so.0()(64bit)
mutter(x86-64) >= 41.0
pipewire-gstreamer(x86-64)
polkit(x86-64) >= 0.100
python3(x86-64)
rpmlib(CompressedFileNames) <= 3.0.4-1
rpmlib(FileDigests) <= 4.6.0-1
rpmlib(PayloadFilesHavePrefix) <= 4.0-1
rpmlib(PayloadIsZstd) <= 5.4.18-1
rpmlib(TildeInVersions) <= 4.10.0-1
rtld(GNU_HASH)
switcheroo-control
upower(x86-64)
xdg-desktop-portal-gtk >= 1.8.0
xdg-user-dirs-gtk
```

File: build/linux/linux-installer/rpm/additional_deps [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+#
+# Additional dependencies not in the rpm find-requires output.
+
+# Make sure users have SSL certificates.
+ca-certificates
+
+# Make sure users have compatible fonts for viewing PDFs.
+liberation-fonts
```
Comment (deepak1556) [inline build/linux/linux-installer/rpm/additional_deps:11]: We don't ship with PDF support.

File: build/linux/linux-installer/rpm/additional_deps [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+# This dependency is added by newer versions of rpmlib.  To ensure consistency
+# of expected dependencies when Chrome is built with the older versions, add it
+# as a manual dependency.
+rpmlib(FileDigests) <= 4.6.0-1
+
+# libvulkan is dlopen()'ed, so must be added as a manual dependency.
+libvulkan.so.1()(64bit)
+
+# For uploading crash reports with Breakpad.
+wget
```
Comment (deepak1556) [inline build/linux/linux-installer/rpm/additional_deps:37]: Breakpad is not used on linux anymore, can be removed.

File: build/linux/rpmDependenciesGenerator.js [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+
+const { spawnSync } = require('child_process');
+const { calculatePackageDeps, mergePackageDeps } = require('./linux-installer/rpm/rpmDependencyScripts');
+const { resolve } = require('path');
+const { readFileSync } = require('fs');
+
+// @ts-check
+function getRpmDependencies() {
+	// Get the files for which we want to find dependencies.
+	const findResult = spawnSync('find', ['.', '-name', '*.node']);
```
Comment (deepak1556) [inline build/linux/rpmDependenciesGenerator.js:16]: Can you run this step on build generated under the `agent.buildDirectory` similar to https://github.com/microsoft/vscode/blob/32d40cf44e893e87ac33ac4f08de1e5f7fe077fc/build/darwin/sign.ts#L29

File: build/linux/linux-installer/rpm/calculate_package_deps.py [hunk 5]
```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python
```
Comment (deepak1556) [inline build/linux/linux-installer/rpm/calculate_package_deps.py:1]: can be removed.

File: build/linux/linux-installer/rpm/merge_package_deps.py [hunk 6]
```diff
@@ -0,0 +1,28 @@
+#!/usr/bin/env python
```
Comment (deepak1556) [inline build/linux/linux-installer/rpm/merge_package_deps.py:1]: can be removed.

File: build/linux/linux-installer/rpm/rpmDependencyScripts.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+const { spawnSync } = require('child_process');
```
Comment (deepak1556) [inline build/linux/linux-installer/rpm/rpmDependencyScripts.js:6]: Also check in the typescript file.

File: build/linux/linux-installer/rpm/update_package_provides.py [hunk 8]
```diff
@@ -0,0 +1,151 @@
+#!/usr/bin/env python
+# Copyright 2017 The Chromium Authors. All rights reserved.
+# Use of this source code is governed by a BSD-style license that can be
+# found in the LICENSE file.
+
+import binascii
```
Comment (deepak1556) [inline build/linux/linux-installer/rpm/update_package_provides.py:6]: What does `--distro-check` do ? Is it helpful for our scenario ?

File: build/linux/rpmDependenciesGenerator.js [hunk 9]
```diff
@@ -0,0 +1,47 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+'use strict';
+
+const { spawnSync } = require('child_process');
```
Comment (deepak1556) [inline build/linux/rpmDependenciesGenerator.js:8]: Also check in the typescript file.

File: build/linux/linux-installer/rpm/rpmDependencyScripts.ts [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+ *  Licensed under the MIT License. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import { spawnSync } from 'child_process';
+import { statSync } from 'fs';
+
+// Based on https://source.chromium.org/chromium/chromium/src/+/main:chrome/installer/linux/rpm/calculate_package_deps.py
+export function calculatePackageDeps(binaryPath: string): Set<string> {
+	try {
+		if ((statSync(binaryPath).mode & 0o111) === 0) {
```
Comment (deepak1556) [inline build/linux/linux-installer/rpm/rpmDependencyScripts.ts:12]:
```suggestion
		if (!(statSync(binaryPath).mode & fs.constants.S_IXUSR)) {
```

File: build/linux/rpmDependenciesGenerator.ts [hunk 11]
```diff
@@ -0,0 +1,41 @@
...
+		return [];
+	}
+
+	const files = findResult.stdout.toString().trimEnd().split('\n');
+
+	const appPath = `${buildDir}/${applicationName}`;
+	files.push(appPath);
+
+	// Generate the dependencies.
+	const dependencies: Set<string>[] = files.map((file) => calculatePackageDeps(file));
```
Comment (deepak1556) [inline build/linux/rpmDependenciesGenerator.ts:27]: Require list should remove the shared libraries the runtime already bundles. for ex: https://source.chromium.org/chromium/chromium/src/+/refs/tags/98.0.4758.109:chrome/installer/linux/BUILD.gn;l=64-80

```
libEGL.so,
libffmpeg.so,
libGLESv2.so,
libvk_swiftshader.so,
libvulkan.so.1
```

these are removed via the `shlibs` argument of `calculate_package_deps.py` 

https://source.chromium.org/chromium/chromium/src/+/refs/tags/98.0.4758.109:chrome/installer/linux/BUILD.gn;l=158
https://source.chromium.org/chromium/chromium/src/+/refs/tags/98.0.4758.109:chrome/installer/linux/rpm/calculate_package_deps.py;l=27
https://source.chromium.org/chromium/chromium/src/+/refs/tags/98.0.4758.109:chrome/installer/linux/rpm/calculate_package_deps.py;l=56-58
https://source.chromium.org/chromium/chromium/src/+/refs/tags/98.0.4758.109:chrome/installer/linux/rpm/calculate_package_deps.py;l=67

Similar filtering can be done here after calculating the dependencies.

File: build/linux/rpmDependenciesGenerator.ts [hunk 12]
```diff
@@ -0,0 +1,41 @@
...
+	if (findResult.status) {
+		console.error('Error finding files:');
+		console.error(findResult.stderr.toString());
+		return [];
+	}
+
+	const files = findResult.stdout.toString().trimEnd().split('\n');
+
+	const appPath = `${buildDir}/${applicationName}`;
+	files.push(appPath);
```
Comment (deepak1556) [inline build/linux/rpmDependenciesGenerator.ts:24]: Should also include `chrome_crashpad_handler` and `chrome-sandbox` executables.

File: build/linux/linux-installer/rpm/rpmDependencyScripts.ts [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+			throw new Error(`Binary ${binaryPath} needs to have an executable bit set.`);
+		}
+	} catch (e) {
+		console.error('Tried to stat ' + binaryPath + ' but failed.');
+		throw e;
+	}
+
+	const findRequiresResult = spawnSync('/usr/lib/rpm/find-requires', [], { input: binaryPath + '\n' });
```
Comment (deepak1556) [inline build/linux/linux-installer/rpm/rpmDependencyScripts.ts:20]:
```suggestion
	const findRequiresResult = spawnSync('/usr/lib/rpm/find-requires', { input: binaryPath + '\n' });
```

File: build/linux/rpmDependenciesGenerator.ts [hunk 14]
```diff
@@ -0,0 +1,41 @@
...
+	const findResult = spawnSync('find', [buildDir, '-name', '*.node']);
+	if (findResult.status) {
+		console.error('Error finding files:');
+		console.error(findResult.stderr.toString());
+		return [];
+	}
+
+	const files = findResult.stdout.toString().trimEnd().split('\n');
+
+	const appPath = `${buildDir}/${applicationName}`;
```
Comment (deepak1556) [inline build/linux/rpmDependenciesGenerator.ts:23]:
```suggestion
	const appPath = path.join(`${buildDir}`, `${applicationName}`);
```

**Review by deepak1556 — COMMENTED** [review]
Looking good mostly, requires some style changes.

File: build/linux/rpm/additional-deps.js [hunk 15]
```diff
@@ -0,0 +1,19 @@
+"use strict";
```
Comment (deepak1556) [inline build/linux/rpm/additional-deps.js:1]: Please remove unused file.

File: build/linux/rpm/dep-lists.ts [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+	'xdg-utils' // OS integration
+];
+
+// Based on https://source.chromium.org/chromium/chromium/src/+/refs/tags/98.0.4758.109:chrome/installer/linux/BUILD.gn;l=64-80
+// and the Linux Archive build
+// Shared library dependencies that we already bundle.
+export const bundledDeps = [
+	'libEGL.so',
+	'libGLESv2.so',
+	'libvulkan.so',
```
Comment (deepak1556) [inline build/linux/rpm/dep-lists.ts:25]: Should be `libvulkan.so.1`

**Review by deepak1556 — COMMENTED** [review]
Can you also remove https://github.com/microsoft/vscode/blob/main/resources/linux/rpm/dependencies.json as part of this PR.

Almost to the finish line, some minor cleanups and should be good to land after build verification.

**Review by deepak1556 — APPROVED** [review]
Verified dependencies change locally

👏

**Pull request merged**